scripts/generate-ffdhe.py

```diff
@@ -0,0 +1,514 @@
+
+		The PEM format consists of a header, the content itself (base64,
+		using the standard alphabet) and a trailer.
+
+		The header is of the form:
+			-----BEGIN $KIND-----
+
+		and the trailer is of the form:
+			-----END $KIND-----
+
+		where $KIND defines the type of data
+		contained in the PEM data.
+
+		Between the header and the trailer is
+		the content itself. The content is the
+		DER form encoded using base64.
+
+		Our implementation splits the base64
+		data such that each line of base64 data
+		can be no longer than 64 characters long.
+	'''
+	out = u'-----BEGIN {0}-----\n'.format(kind)
+	enc = base64.b64encode(der)
+	n = 0
+	for ch in enc:
+		if type(ch) == int: # Python 3
+			out += chr(ch)
+		else:
+			out += ch
+		n += 1
+		if n == 64:
+			out += '\n'
+			n = 0
+	if out[-1] != '\n':
+		out += '\n'
+	out += u'-----END {0}-----\n'.format(kind)
+	return out
+
+def writeHeader(f):
+	'''
+		Output the LICENSE.header file plus a warning
+		about the output file being auto-generated.
+
+	'''
+	hf = codecs.open('LICENSE.header', 'r', 'utf-8')
+	hdr = hf.read()
+	hf.close()
+
+	f.write(hdr)
+	f.write('\n')
+	f.write('// Automatically generated by "generate-ffdhe.py". DO NOT EDIT BY HAND.\n')
+	f.write('//\n')
+	f.write('// I also agree to have manually vetted this file for correctness.\n')
+	f.write('//\n')
+	f.write('// If I do not agree, I will not have removed the line above saying\n')
+	f.write('// otherwise. Nor will I have removed the line below this one which\n')
+	f.write('// will cause a preprocessor error. Oops!\n')
+	f.write('#error Please verify this file is correct\n')
+	f.write('\n')
+
+def writeCStr(f, identifier, str):
+	'''
+		Output the string 'str' as a static
+		C-style string (in array form, so sizeof works).
+	'''
+	f.write('static const char {0}[] = \\\n'.format(identifier))
+	splat = str.split('\n')
+	splat = splat[:-1]
+	lastLine = splat[-1]
+	for line in splat:
+		terminator = ''
+		if line == lastLine:
+			terminator = ';'
+		if len(line) > 0:
+			f.write('"{0}\\n"{1}\n'.format(line, terminator))
+def main():
+	f = codecs.open(os.path.join('src', 'FFDHETable.h'), 'w', 'utf-8')
+
+	writeHeader(f)
+
+	writeCStr(f, 'ffdhe2048_pem', pem(
+		derSequence((
+			derUnsignedInteger(trim(ffdhe2048_p)),
+			derUnsignedInteger(trim(g)),
+		))
+	))
+
+	f.write('\n')
+
+	writeCStr(f, 'ffdhe3072_pem', pem(
+		derSequence((
+			derUnsignedInteger(trim(ffdhe3072_p)),
+			derUnsignedInteger(trim(g)),
+		))
+	))
+
+	f.write('\n')
+
+	writeCStr(f, 'ffdhe4096_pem', pem(
+		derSequence((
+			derUnsignedInteger(trim(ffdhe4096_p)),
+			derUnsignedInteger(trim(g)),
+		))
+	))
+
+	f.write('\n')
+
+	writeCStr(f, 'ffdhe6144_pem', pem(
+		derSequence((
+			derUnsignedInteger(trim(ffdhe6144_p)),
+			derUnsignedInteger(trim(g)),
+		))
+	))
+
+	f.write('\n')
+
+	writeCStr(f, 'ffdhe8192_pem', pem(
+		derSequence((
+			derUnsignedInteger(trim(ffdhe8192_p)),
+			derUnsignedInteger(trim(g)),
+		))
+	))
```

Comment:
Redundant block structure could be put into a function with parameter `f`, `ffdhePamName`, `p`, `g`.